File: scripts/test_pattern/test_decorator.py
"""Tests for the Decorator pattern."""

import sys
import os
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'pattern'))
from decorator import (
    Espresso, SimpleCoffee,
    MilkDecorator, CaramelDecorator, WhipDecorator
)


class TestDecorator(unittest.TestCase):

    EPS = 1e-9

    # ------------------------------------------------------------------
    # Base component
    # ------------------------------------------------------------------
    def test_espresso_cost(self):
        self.assertAlmostEqual(1.00, Espresso().cost())

    def test_espresso_description(self):
        self.assertEqual("Espresso", Espresso().description())

    # ------------------------------------------------------------------
    # Single decorator
    # ------------------------------------------------------------------
    def test_milk_adds_cost(self):
        self.assertAlmostEqual(1.25, MilkDecorator(Espresso()).cost())

    def test_milk_adds_description(self):
        desc = MilkDecorator(Espresso()).description()
        self.assertIn("Milk",     desc)
        self.assertIn("Espresso", desc)

    def test_caramel_adds_cost(self):
        self.assertAlmostEqual(1.50, CaramelDecorator(Espresso()).cost())

    # ------------------------------------------------------------------
    # Stacked decorators
    # ------------------------------------------------------------------
    def test_milk_and_caramel_cost(self):
        c = CaramelDecorator(MilkDecorator(Espresso()))
        self.assertAlmostEqual(1.75, c.cost())

    def test_triple_stack_cost(self):
        c = WhipDecorator(CaramelDecorator(MilkDecorator(Espresso())))
        self.assertAlmostEqual(2.05, c.cost())

    def test_triple_stack_description(self):
        c = WhipDecorator(CaramelDecorator(MilkDecorator(Espresso())))
        desc = c.description()
        for word in ("Espresso", "Milk", "Caramel", "Whip"):
            self.assertIn(word, desc)

    # ------------------------------------------------------------------
    # Same decorator twice
    # ------------------------------------------------------------------
    def test_double_milk(self):
        c = MilkDecorator(MilkDecorator(Espresso()))
        self.assertAlmostEqual(1.50, c.cost())

    # ------------------------------------------------------------------
    # Polymorphic use
    # ------------------------------------------------------------------
    def test_simple_coffee_with_milk(self):
        self.assertAlmostEqual(0.75, MilkDecorator(SimpleCoffee()).cost())


if __name__ == "__main__":
    unittest.main()

File: test/pattern/test_decorator.cpp
#include "pattern/decorator.hpp"
#include <cmath>
#include <gtest/gtest.h>
#include <memory>

using namespace pattern;

static const double EPS = 1e-9;

// ---------------------------------------------------------------------------
// Base component
// ---------------------------------------------------------------------------
TEST(Decorator, EspressoBaseCost)
{
    Espresso e;
    EXPECT_NEAR(1.00, e.cost(), EPS);
}

TEST(Decorator, EspressoDescription)
{
    Espresso e;
    EXPECT_EQ("Espresso", e.description());
}

// ---------------------------------------------------------------------------
// Single decorator
// ---------------------------------------------------------------------------
TEST(Decorator, MilkAddsCost)
{
    std::unique_ptr<Coffee> c(new MilkDecorator(std::unique_ptr<Coffee>(new Espresso())));
    EXPECT_NEAR(1.25, c->cost(), EPS);
}

TEST(Decorator, MilkAddsDescription)
{
    std::unique_ptr<Coffee> c(new MilkDecorator(std::unique_ptr<Coffee>(new Espresso())));
    EXPECT_NE(std::string::npos, c->description().find("Milk"));
    EXPECT_NE(std::string::npos, c->description().find("Espresso"));
}

TEST(Decorator, CaramelAddsCost)
{
    std::unique_ptr<Coffee> c(new CaramelDecorator(std::unique_ptr<Coffee>(new Espresso())));
    EXPECT_NEAR(1.50, c->cost(), EPS);
}

// ---------------------------------------------------------------------------
// Stacked decorators
// ---------------------------------------------------------------------------
TEST(Decorator, MilkAndCaramelCost)
{
    std::unique_ptr<Coffee> c = std::unique_ptr<Coffee>(
        new CaramelDecorator(std::unique_ptr<Coffee>(new MilkDecorator(std::unique_ptr<Coffee>(new Espresso())))));
    EXPECT_NEAR(1.75, c->cost(), EPS);
}

TEST(Decorator, TripleStackDescription)
{
    std::unique_ptr<Coffee> c    = std::unique_ptr<Coffee>(new WhipDecorator(
        std::unique_ptr<Coffee>(new CaramelDecorator(
            std::unique_ptr<Coffee>(new MilkDecorator(std::unique_ptr<Coffee>(new Espresso())))))));
    const std::string       desc = c->description();
    EXPECT_NE(std::string::npos, desc.find("Espresso"));
    EXPECT_NE(std::string::npos, desc.find("Milk"));
    EXPECT_NE(std::string::npos, desc.find("Caramel"));
    EXPECT_NE(std::string::npos, desc.find("Whip"));
    EXPECT_NEAR(2.05, c->cost(), EPS);
}

// ---------------------------------------------------------------------------
// Same decorator applied twice
// ---------------------------------------------------------------------------
TEST(Decorator, DoubleMilk)
{
    std::unique_ptr<Coffee> c = std::unique_ptr<Coffee>(
        new MilkDecorator(std::unique_ptr<Coffee>(new MilkDecorator(std::unique_ptr<Coffee>(new Espresso())))));
    EXPECT_NEAR(1.50, c->cost(), EPS);
}

// ---------------------------------------------------------------------------
// Polymorphic usage via base pointer
// ---------------------------------------------------------------------------
TEST(Decorator, PolymorphicDecoration)
{
    std::unique_ptr<Coffee> base(new SimpleCoffee());
    std::unique_ptr<Coffee> decorated(new MilkDecorator(std::move(base)));
    EXPECT_NEAR(0.75, decorated->cost(), EPS);
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
